src/QsCompiler/Compiler/Templates/CppInterop.cs

```diff
@@ -0,0 +1,168 @@
+            this.Type = argument.Type;
+            this.ArrayType = argument.ArrayType;
+            this.Values = argument.Values.ToList();
+        }
+
+        public string CppType()
+        {
+            return this.Type switch
+            {
+                DataType.BoolType => "char",
+                DataType.IntegerType => "int64_t",
+                DataType.DoubleType => "double",
+                DataType.PauliType => "char",
+                DataType.RangeType => "InteropRange*",
+                DataType.ResultType => "char",
+                DataType.StringType => "const char*",
+                DataType.ArrayType => "InteropArray *",
+            };
+        }
+
+        public string CliOptionString()
+        {
+            if (this.Name.Length == 1)
+            {
+                return $"-{this.Name}";
+            }
+            else
+            {
+                return $"--{this.Name}";
+            }
+        }
+
+        public string? CliDescription()
+        {
+            return this.Type switch
+            {
+                DataType.BoolType => $"A bool value for the {this.Name} argument",
+                DataType.IntegerType => $"An integer value for the {this.Name} argument",
+                DataType.DoubleType => $"A double value for the {this.Name} argument",
+                DataType.PauliType => $"A Pauli value for the {this.Name} argument",
+                DataType.RangeType => $"A Range value (start, step, end) for the {this.Name} argument",
+                DataType.ResultType => $"A Result value for the {this.Name} argument",
+                DataType.StringType => $"A String value for the {this.Name} argument",
+                DataType.ArrayType => this.ArrayType switch
+                {
+                    DataType.BoolType => $"A bool array for the {this.Name} argument",
+                    DataType.IntegerType => $"An integer array for the {this.Name} argument",
+                    DataType.DoubleType => $"A double array for the {this.Name} argument",
+                    DataType.PauliType => $"A Pauli array for the {this.Name} argument",
+                    DataType.RangeType => $"A Range array for the {this.Name} argument",
+                    DataType.ResultType => $"A Result array for the {this.Name} argument",
+                    DataType.StringType => $"A String array for the {this.Name} argument",
+                },
+            };
+        }
+
+        public string? CppVarType()
+        {
+            return this.Type switch
+            {
+                DataType.BoolType => "char",
+                DataType.IntegerType => "int64_t",
+                DataType.DoubleType => "double_t",
+                DataType.PauliType => "PauliId",
+                DataType.RangeType => "RangeTuple",
+                DataType.ResultType => "char",
+                DataType.StringType => "string",
+                DataType.ArrayType => this.ArrayType switch
```

Comment:
For all switch expressions, can you please throw an exception similar to the one in this [comment](https://github.com/microsoft/qsharp-compiler/pull/929/files#r609269397) when the argument type is not explicitly supported.

Reply:
Also the for the sub-expressions that switch on `ArrayType`.

Reply:
done